### PR TITLE
[CDF-25543] 📖CSV schema sniffer

### PR DESCRIPTION
**Pull request by doctrino**
# Description

<del>Stacked #1855, #1856 </del>

**Context**: We have a `csv` reader that naively converts cell values with custom logic. Instead we want to sniff the schema from the file and ensure that all columns have the data type. 

This PR adds a sniffer to the CSV file reader

## Changelog

- [ ] Patch
- [x] Skip

### CODE REVIEW

**Review by muradsater — APPROVED** [review]

**Review by muradsater — APPROVED** [review]

**Comment by thorkildcognite** [comment]
@doctrino @muradsater Due to the automatic rebase, this needs a new team review!

File: cognite_toolkit/_cdf_tk/utils/fileio/_readers.py [hunk 1]
```diff
@@ -75,34 +79,130 @@ class YMLReader(YAMLBaseReader):
...
-        return value
+        return infer_data_type_from_value(value, "Json")[1]
+
+    @classmethod
+    def sniff_schema(cls, input_file: Path, sniff_rows: int) -> list[SchemaColumn]:
+        """Sniff the schema from the first `sniff_rows` rows of the CSV file."""
+        if sniff_rows <= 0:
+            raise ValueError("`sniff_rows` must be a positive integer.")
+
+        with input_file.open("r", encoding="utf-8") as file:
```
Comment (rsjr) [inline cognite_toolkit/_cdf_tk/utils/fileio/_readers.py:155]: Maybe I got this whole thing wrong, but are you doing file open operation twice here? I mean, when you are going instantiate CSVReader, you are opening the file, and then reopening it again here to "sniff the schema"?

File: cognite_toolkit/_cdf_tk/utils/fileio/_readers.py [hunk 2]
```diff
@@ -75,34 +79,130 @@ class YMLReader(YAMLBaseReader):
...
+            for key, value in row.items():
+                if value == "":
+                    parsed[key] = None
+                    continue
+                try:
+                    parsed[key] = self.parse_function_by_column[key](value)
+                except ValueError as e:
+                    parsed[key] = None
+                    if self.keep_failed_cells:
+                        self.failed_cell.append(FailedParsing(row=row_no, column=key, value=value, error=str(e)))
```
Comment (rsjr) [inline cognite_toolkit/_cdf_tk/utils/fileio/_readers.py:204]: failed_cell is not cleared anywhere, so the errors are going to get appended everytime reader is called?

File: tests/test_unit/test_cdf_tk/test_utils/test_fileio.py [hunk 3]
```diff
@@ -322,6 +323,62 @@ def part_number(filepath: Path) -> int:
 
 
 class TestCSVReader:
+    CSV_CONTENT = """text,integer,nested,boolean,float
```
Comment (rsjr) [inline tests/test_unit/test_cdf_tk/test_utils/test_fileio.py:326]: Can you add some edge cases here such as:
- Empty file
- File cointaining only header
- File with duplicate names in the column header

**Review by rsjr — COMMENTED** [review]
Left some comments. Also, a question regarding performance when reading the CSV files. How big are the files you are going to read here? Have you done any load tests (manually is fine)?

File: tests/test_unit/test_cdf_tk/test_utils/test_fileio.py (hunk 3 above)
Comment (doctrino) [reply tests/test_unit/test_cdf_tk/test_utils/test_fileio.py:326]: I added that and a few more cases as through usage of this (in upcoming PRs) it is very natural to start with calling this method. Then, it is good that it handles file not exists and so on.

**Review by Aashutosh-cognite — APPROVED** [review]

**Review by rsjr — COMMENTED** [review]
risk review ok
